### ISSUE TITLE
kv hash key partitioning

### ISSUE THREAD

**Issue by nopcoder** [treeverse/lakeFS#3399]
Enabling better underlaying performance can be achieved by adding another level, we can partition the data.
Postgres table can be partitioned by a key - to enable that, we can add an optional hash-key while setting and scanning data in our kv storage.

This issue was created to address:
- API changes to enable operations using hash-key
- Update the Postgres implementation to support partitioning

**Comment by itaiad200** [comment]
Putting in writing the discussion @nopcoder And I had with @ozkatz:

The API should not be changed. It is written in a general way on purpose to allow many implementations to the kv store interface. We want to partition on Postgres since the current huge non-partitioned table will quickly become the bottleneck. 
The partitioning must take into account the main entity type with many changes and where performance matters => entries under a staging token. Since Keys are opaque to the kv store, it must decide on a parameter to partition by - which most likely be the prefix of the key. To ensure all entries under a staging token ends up in the same partition, we must write it at the very beginning of the key. e.g. instead of `refstore/repo/my-repo/branch/my-branch/stagingTokens/my-staging-token/path/to/my/file` do `stagingTokens/my-staging-token/repo/my-repo/branch/my-branch/path/to/my/file/`.

Given that we're currently not facing this problem (only multipart implementation is completed), and assuming that we agree on the above, this becomes more of a principal than an actual task to be done. I'll close this one if no objections are raised.

**Comment by arielshaqed** [comment]
I pretty much disagree with [all of the above](https://github.com/treeverse/lakeFS/issues/3399#issuecomment-1134552176)

The API without separate hash and ordered keys is undoubtedly the simplest.  However not by much, and OTOH it will be very hard to scale.

First, note that _almost every key needs a prefix space_.  By not supporting it in the KV API, you're just moving the problem of handling prefixes over to the application.  Which is actually a harder problem than it seems if you don't want to limit the bytes allowed in your prefix.  So simplest might be to decide to go with `prefix`**:**`key`, but that means you don't get to use colons in your prefix.  

But more importantly, the current API does not match some useful real-life systems, and will not attain full performance.  Examples of high-value implementation targets:

* DynamoDB requires a hash and an ordered key.  Taking the first few bytes of the key as a hash will only work if `scan` becomes a complex operation.  Failing that, it is not that nice.
* PostgreSQL is not that great at dropping many keys.  This is precisely what we need to do during a COMMIT.  But if all staging keys live in the same partitions (or even just a separate table!), this is as simple as DROPping the partition table (and faster because the staging area partition is uncontended by the time you drop it).
* Cassandra and ScyllaDB have the [same type of split key](https://university.scylladb.com/courses/data-modeling/lessons/basic-data-modeling-2/topic/primary-key-partition-key-clustering-key/) as DynamoDB.  True, you might want to split the different staging areas internally to avoid hotspotting -- but at the very least the API has to acknowledge this.

So I am not sold on this.  It is a simpler API.  But that API prevents a trivial performance boost on DynamoDB and similar KV stores, and will require a difficult schema upgrade when we do decide to move.  And while the API is simpler, the actual implementation code needs to be smarter about how it handles partitions.  Given that actual application code _**anyway**_ needs to provide a keyspace (because auth != staging areas != ...), this simpler API makes the application code harder to migrate to real databases.

**Comment by ozkatz** [comment]
OK I think I _now_ understand what you mean by "hash key" and it's actually a partition key? a-la dynamodb/cassandra?
I'd first suggest naming it "partition key".

> Cassandra and ScyllaDB have the [same type of split key](https://university.scylladb.com/courses/data-modeling/lessons/basic-data-modeling-2/topic/primary-key-partition-key-clustering-key/) as DynamoDB.

 I'm ok with the split key concept, I do think it's a rather short blanket anyway. 

If we want to [support databases](https://github.com/treeverse/lakeFS/blob/master/design/open/metadata_kv/index.md#databases-that-meet-these-requirements-examples) that don't have this construct it does become harder: object stores, foundationDB, spanner/cockroach and possibly others - don't provide this type of control and partition automatically on a contiguously sorted key space for scaling throughput (see S3's now-outdated recommendation to prefix paths with hash values). 

Heck, this is even what lakeFS does :)


> PostgreSQL is not that great at dropping many keys

'm not as concerned about performance there since this should not block a commit request anyway. dropping a partition isn't good enough unless we tell Postgres to match its partition count to the cardinality of e.g. staging tokens which could potentially mess up point lookup time (as most accesses will likely hit a partition with still-empty stats, resulting in ~table~ partition scans). 

This requires testing but we've been burned by this sort of unpredictability before and I think we can agree that critical path random lookups should be prioritized over efficient partition delete performance.

**TL;DR** - I'm OK with having a partition key in the API. I'd be careful about premature optimizations in non-critical-path areas.
